### PR TITLE
Convenience API and bug fixes for git tree walking

### PR DESCRIPTION
**Pull request by maleadt**
This PR fixes a bug with LibGit2's `treewalk` wrapper, and adds a convenience `getindex`-based API for accessing objects in a tree by path. First time working with libgit2(.jl), so feel free to bikeshed 🙂

EDIT: I only ran the `libgit2` tests locally because of the device I'm working on, so this might cause test failures.

### CODE REVIEW

File: base/libgit2/oid.jl
Comment (maleadt) [inline base/libgit2/oid.jl]: This seemed like a reasonable thing to do, but is not really required for this PR (it simplifies a test).

File: base/libgit2/tree.jl
Comment (ararslan) [inline base/libgit2/tree.jl]: Base still uses `->` to show return types in docstrings

File: base/libgit2/oid.jl
Comment (ararslan) [reply base/libgit2/oid.jl]: It seems reasonable to me as well, so long as there's a test for it. :slightly_smiling_face:

File: base/libgit2/tree.jl
Comment (bramtayl) [reply base/libgit2/tree.jl]: why?

Comment (StefanKarpinski) [reply base/libgit2/tree.jl]: Partly historical artifact, partly because it's often used in such a way that the `:: T` return type syntax would be wrong or inappropriate.

**Comment by maleadt** [comment]
Bump? Comments have been addressed. Or should this wait until after 1.0?

**Comment by StefanKarpinski** [comment]
No, this can be merged now if someone familiar with the LibGit2 code takes a look at it. Maybe @kshyatt?

File: base/libgit2/tree.jl
Comment (kshyatt) [inline base/libgit2/tree.jl]: X-refs here?

**Review by kshyatt — APPROVED** [review]
other than the minor request to xref docs this lgtm sorry for the wait on the review

**Comment by maleadt** [comment]
OK, addressed all comments. Will merge if CI is green.

**Comment by StefanKarpinski** [comment]
AppVeyor failures where libgit2 remote exceptions. I've restarted them to make sure they're not actually related since this change does touch libgit2 code. Suspiciously, the Travis Linux 64 failure was _also_ a libgit2 remote exception. I've restarted that one as well, so we'll see...

**Comment by maleadt** [comment]
I'll look into it if any more suspicious failures turn up, after all I've also rebased the branch which might have introduced something.

**Comment by maleadt** [comment]
So this keeps throwing the `RemoteException` on AppVeyor during libgit2 tests, while other CI services are doing fine (only CircleCI fails, but due to a different ProcessExitedException). So this is probably related, I guess? Annoying how I can't see the original exception though (I assume this is a known issue?):

```
libgit2: Error During Test at nothing:1
  Test threw an exception of type RemoteException
  Expression: libgit2
  Error deserializing a remote exception from worker 11
  Remote(original) exception of type TestSetException
  Remote stacktrace : 
  finish at C:\projects\julia\usr\share\julia\site\v0.7\Test\src\Test.jl:784
  macro expansion at .\util.jl:310 [inlined]
  top-level scope at .\<missing>:19
  eval at .\boot.jl:295
  #runtests#1 at C:\projects\julia\julia-31f138a0c1\share\julia\test\testdefs.jl:25
  #runtests at .\<missing>:0 [inlined] (repeats 2 times)
  #125 at C:\projects\julia\usr\share\julia\site\v0.7\Distributed\src\process_messages.jl:269
  run_work_thunk at C:\projects\julia\usr\share\julia\site\v0.7\Distributed\src\process_messages.jl:56
  macro expansion at C:\projects\julia\usr\share\julia\site\v0.7\Distributed\src\process_messages.jl:269 [inlined]
```

I'll set-up a windows system tomorrow and debug this.

**Comment by simonbyrne** [comment]
You can also RDP to an appveyor job: https://www.appveyor.com/docs/how-to/rdp-to-build-worker/

**Comment by maleadt** [comment]
OK, this was a bug on Windows indeed: I was using `joinpath`, while libgit2 expects Unix path separators. Didn't know debugging Appveyor was such a pain though; the stacktrace was clear (ie. no RemoteException) after RDP'ing into the builder.

Only remaining CI failure is a seemingly unrelated ProcessExitedException on CircleCI; @KristofferC is it OK if I merge this before #25706?

**Comment by KristofferC** [comment]
> @KristofferC is it OK if I merge this before #25706?

I've rebased #25706 N times, so rebasing it N+1 times is no problem.

**Pull request merged**
